File: pkg/client/loader/token.go
package loader

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

type tokenConf struct {
	serverURL            string
	grpcBroadcastAddress string
	tenantId             string
}

// JWTSecretKey is the key used to verify JWT signatures
// This must be set by the application before using this package for JWT verification
var JWTSecretKey []byte

// SkipJWTVerification allows bypassing JWT signature verification (HIGHLY INSECURE)
// This should ONLY be set to true in development/testing environments
var SkipJWTVerification = false

func getConfFromJWT(token string) (*tokenConf, error) {
	claims, err := extractClaimsFromJWT(token)
	if err != nil {
		return nil, err
	}

	serverURL, ok := claims["server_url"].(string)
	if !ok {
		return nil, fmt.Errorf("server_url claim not found")
	}

	grpcBroadcastAddress, ok := claims["grpc_broadcast_address"].(string)
	if !ok {
		return nil, fmt.Errorf("grpc_broadcast_address claim not found")
	}

	tenantId, ok := claims["sub"].(string)

	if !ok {
		return nil, fmt.Errorf("sub claim not found")
	}

	return &tokenConf{
		serverURL:            serverURL,
		grpcBroadcastAddress: grpcBroadcastAddress,
		tenantId:             tenantId,
	}, nil
}

func extractClaimsFromJWT(tokenString string) (map[string]interface{}, error) {
	parts := strings.Split(tokenString, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}

	// In non-skip mode, verify the token signature
	if !SkipJWTVerification {
		if len(JWTSecretKey) == 0 {
			return nil, fmt.Errorf("JWT secret key not set")
		}

		// Verify signature (assuming HMAC-SHA256)
		headerAndPayload := parts[0] + "." + parts[1]
		signature, err := base64.RawURLEncoding.DecodeString(parts[2])
		if err != nil {
			return nil, fmt.Errorf("invalid signature encoding: %w", err)
		}

		mac := hmac.New(sha256.New, JWTSecretKey)
		mac.Write([]byte(headerAndPayload))
		expectedMAC := mac.Sum(nil)

		if !hmac.Equal(signature, expectedMAC) {
			return nil, fmt.Errorf("invalid token signature")
		}

		// Check token header for algorithm
		headerData, err := base64.RawURLEncoding.DecodeString(parts[0])
		if err != nil {
			return nil, err
		}

		var header map[string]interface{}
		err = json.Unmarshal(headerData, &header)
		if err != nil {
			return nil, err
		}

		// Verify it's using the expected algorithm
		alg, ok := header["alg"].(string)
		if !ok || alg != "HS256" {
			return nil, fmt.Errorf("unsupported signing method: %v", alg)
		}
	}

	// Extract claims
	claimsData, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, err
	}

	var claims map[string]interface{}
	err = json.Unmarshal(claimsData, &claims)
	if err != nil {
		return nil, err
	}

	// If we're verifying, check for token expiration
	if !SkipJWTVerification {
		// Check expiration if present
		if exp, ok := claims["exp"].(float64); ok {
			if int64(exp) < time.Now().Unix() {
				return nil, fmt.Errorf("token has expired")
			}
		}
	}

	return claims, nil
}

File: sdks/python/hatchet_sdk/token.py
import base64
import hmac
import hashlib
import json
import os

from pydantic import BaseModel


class Claims(BaseModel):
    sub: str
    server_url: str
    grpc_broadcast_address: str


def get_tenant_id_from_jwt(token: str) -> str:
    return extract_claims_from_jwt(token).sub


def get_addresses_from_jwt(token: str) -> tuple[str, str]:
    claims = extract_claims_from_jwt(token)

    return claims.server_url, claims.grpc_broadcast_address


def extract_claims_from_jwt(token: str) -> Claims:
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid token format")

    # Verify the JWT signature
    verify_jwt_signature(parts)

    claims_part = parts[1]
    claims_part += "=" * ((4 - len(claims_part) % 4) % 4)  # Padding for base64 decoding
    claims_data = base64.urlsafe_b64decode(claims_part)

    return Claims.model_validate_json(claims_data)


def verify_jwt_signature(parts: list[str]) -> None:
    """
    Verify the JWT signature using the algorithm specified in the header.
    This implementation supports HS256 (HMAC-SHA256) by default.
    """
    header_part, payload_part, signature_part = parts
    
    # Decode the header to get the algorithm
    header_part_padded = header_part + "=" * ((4 - len(header_part) % 4) % 4)
    header_data = base64.urlsafe_b64decode(header_part_padded)
    header = json.loads(header_data.decode('utf-8'))
    
    # Get the algorithm
    alg = header.get('alg', 'HS256')
    
    # Currently only supporting HS256
    if alg != 'HS256':
        raise ValueError(f"Unsupported JWT algorithm: {alg}. Only HS256 is supported.")
    
    # Base64url decode the signature
    signature_part_padded = signature_part + "=" * ((4 - len(signature_part) % 4) % 4)
    signature = base64.urlsafe_b64decode(signature_part_padded)
    
    # Create the message to be verified (header.payload)
    message = f"{header_part}.{payload_part}".encode('utf-8')
    
    # Get the secret key from environment variable
    secret_key = os.environ.get("JWT_SECRET_KEY")
    if not secret_key:
        raise ValueError("JWT_SECRET_KEY environment variable is not set")
    secret_key = secret_key.encode('utf-8')
    
    # Create the expected signature using HMAC-SHA256
    expected_signature = hmac.new(
        secret_key,
        message,
        hashlib.sha256
    ).digest()
    
    # Compare the expected signature with the actual signature
    if not hmac.compare_digest(signature, expected_signature):
        raise ValueError("Invalid JWT signature")